DFT/jesus_dft.py

```diff
@@ -574,3 +575,340 @@ def show_signal_cosines(self):
         self.play(Write(original_freq_mob))
         self.play(FadeIn(times))
         self.play(LaggedStartMap(FadeIn, analysis_freqs), run_time=2)
+
+
+class IntroducePhaseProblem(MovingCameraScene):
+    def construct(self):
+
+        frame = self.camera.frame.save_state()
+        self.try_sine_wave()
+
+        self.play(*[FadeOut(mob) for mob in self.mobjects])
+        self.play(Restore(frame))
+
+        self.test_cases_again()
+
+    def try_sine_wave(self):
+        frame = self.camera.frame
+        t_min = 0
+        t_max = TAU
+
+        original_freq = 2
+
+        # samples per second
+        sample_frequency = 80
+
+        n_samples = sample_frequency
+
+        vt_frequency = ValueTracker(original_freq)
+        vt_phase = ValueTracker(0)
+        vt_amplitude = ValueTracker(1)
+        vt_b = ValueTracker(0)
+
+        def sine_cosine_redraw():
+            phase_ch_cos = get_cosine_func(
+                amplitude=vt_amplitude.get_value(),
+                freq=vt_frequency.get_value(),
+                phase=vt_phase.get_value(),
+                b=vt_b.get_value(),
+            )
+            _, phase_ch_cos_mob = plot_time_domain(phase_ch_cos, t_max=t_max)
+
+            return phase_ch_cos_mob.scale(0.6).shift(UP * 1.5)
+
+        af_matrix = get_analysis_frequency_matrix(
+            N=n_samples, sample_rate=sample_frequency, t_max=t_max
+        )
+
+        rect_scale = 0.1
+
+        def updating_transform_redraw():
+            signal_function = get_cosine_func(
+                amplitude=vt_amplitude.get_value(),
+                freq=vt_frequency.get_value(),
+                phase=vt_phase.get_value(),
+                b=vt_b.get_value(),
+            )
+
+            sampled_signal = np.array(
+                [
+                    signal_function(v)
+                    for v in np.linspace(t_min, t_max, num=n_samples, endpoint=False)
+                ]
+            ).reshape(-1, 1)
+
+            # matrix transform
+            mt = apply_matrix_transform(sampled_signal, af_matrix)
+
+            rects = (
+                VGroup(
+                    *[
+                        VGroup(
+                            Rectangle(
+                                color=REDUCIBLE_VIOLET, width=0.3, height=f * rect_scale
+                            ).set_fill(REDUCIBLE_VIOLET, opacity=1),
+                            Text(str(i), font=REDUCIBLE_MONO).scale(0.4),
+                        ).arrange(DOWN)
+                        for i, f in enumerate(mt.flatten()[: mt.shape[0] // 2])
```

Comment:
Double check this, but I believe this might be a bug after some further thought. 

The `get_analysis_frequency_matrix` function already cuts the dimension of the rows to half of the number of samples N, so if you do this again, you will actually cut the number of plotted frequencies to a quarter of the number of original samples, when in reality, we should only be cutting it in half. 

One thing that lead me to notice this is you were getting 20 samples, even though you originally gave it 80 points.

Reply:
Here's a suggested function that implements this that I am using with success so far:

```
def get_fourier_rects(
      signal_func,
      n_samples=32,
      sample_rate=32,
      t_min=0,
      t_max=2 * PI,
      rect_scale=0.1,
  ):
      # idea: sample points from get_fourier_line_chart
      # and create num_bars dynamically to scale with the axes size
      af_matrix = get_analysis_frequency_matrix(
          N=n_samples, sample_rate=sample_rate, t_max=t_max
      )
      print(af_matrix.shape)
      sampled_signal = np.array(
          [
              signal_func(v)
              for v in np.linspace(t_min, t_max, num=n_samples, endpoint=False)
          ]
      ).reshape(-1, 1)
      mt = apply_matrix_transform(sampled_signal, af_matrix)
      rects = (
          VGroup(
              *[
                  VGroup(
                      Rectangle(
                          color=REDUCIBLE_VIOLET, width=0.3, height=f * rect_scale
                      ).set_fill(REDUCIBLE_VIOLET, opacity=1),
                      Text(str(i), font=REDUCIBLE_MONO).scale(0.4),
                  ).arrange(DOWN)
                  for i, f in enumerate(mt.flatten()[: mt.shape[0]])
              ]
          )
          .arrange(RIGHT, aligned_edge=DOWN)
          .scale(0.6)
      )

    return rects
```

Reply:
You are totally right about this, but it was giving me problems before. I tried your solution and works as expected, though I remember it not working properly without the ```mt.shape[0] // 2```. But it was probably a problem of something else and not that, that I was not aware I resolved. 

Thanks!